src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMBase.scala

```diff
@@ -39,10 +41,38 @@ trait LightGBMBase[TrainedModel <: Model[TrainedModel]] extends Estimator[Traine
     }
   }
 
-  protected def prepareDataframe(dataset: Dataset[_], trainingCols: ListBuffer[String],
+  protected def castColumns(dataset: Dataset[_], trainingCols: ListBuffer[(String, DataType)],
+                            selectedCols: ListBuffer[Column]): DataFrame = {
+    val schema = dataset.schema
+    // Figure out which columns need to be cast
+    val castCols = trainingCols.flatMap {
```

Comment:
no need for two loops here as they are the same collections

Reply:
great suggestion, done!